### PR TITLE
🐛 Removes MachineSet Controller owner for existing Machines

### LINKED ISSUE #3145: v1alpha3 Machine controller unable to process Machines created from v1alpha2 MachineSets
**What steps did you take and what happened:**
1. Deploy v1alpha2 and create a Cluster + MachineSet
1. Upgrade the management cluster to v1alpha3
1. Machine controller has errors such as 
```
E0604 21:54:31.920616       1 controller.go:258] controller-runtime/controller "msg"="Reconciler error" "error"="[Object namespace1/make-20-worker-7dszw is already owned by another MachineSet controller make-20-worker-77c4b8d5f, Object namespace1/make-20-worker-59wn9 is already owned by another MachineSet controller make-20-worker-77c4b8d5f]"  "controller"="machine" "request"={"Namespace":"namespace1","Name":"make-20-worker-77c4b8d5f-fhvrc"}
```

**What did you expect to happen:**
1. No errors

**Anything else you would like to add:**
In v1alpha2, the MachineSet controller clones the infra machine and bootstrap config templates and sets the MachineSet as a controller owner.

In v1alpha3, the Machine controller tries to set the Machine as the controller owner for these resources, and it's not allowed because the MachineSet is already the controller owner.

We probably need to write some logic in the Machine controller to remove any MachineSet controller owner that it finds, as we really want the Machine to be the owner.

**Environment:**

- Cluster-api version: 0.3.6

/kind bug

### PR DESCRIPTION
**Pull request by srm09**
**What this PR does / why we need it**:
Checks if any existing machines have the MachineSet controllers as the owner and removes them in the reconcile phase since the Machine controller should be the owner

**Which issue(s) this PR fixes**:
Fixes #3145

### CODE REVIEW

File: controllers/machine_controller_phases.go [hunk 1]
```diff
@@ -114,6 +114,18 @@ func (r *MachineReconciler) reconcileExternal(ctx context.Context, cluster *clus
 		return external.ReconcileOutput{}, err
 	}
 
+	// remove machineset controller onwer, if any since the Machine should be the owner.
+	controller := metav1.GetControllerOf(obj)
+	if controller != nil {
```
Comment (vincepri) [inline controllers/machine_controller_phases.go:119]:
```suggestion
	if controller := metav1.GetControllerOf(obj); controller != nil {
```

File: util/util.go [hunk 2]
```diff
@@ -385,6 +385,13 @@ func ReplaceOwnerRef(ownerReferences []metav1.OwnerReference, source metav1.Obje
 	return ownerReferences
 }
 
+func RemoveOwnerRef(ownerReferences []metav1.OwnerReference, source metav1.OwnerReference) []metav1.OwnerReference {
```
Comment (vincepri) [inline util/util.go:388]: Add godoc

File: controllers/machine_controller_phases.go (hunk 1 above)
Comment (srm09) [reply controllers/machine_controller_phases.go]: Done

File: util/util.go (hunk 2 above)
Comment (srm09) [reply util/util.go]: Done

File: controllers/machine_controller_phases.go [hunk 3]
```diff
@@ -114,6 +114,17 @@ func (r *MachineReconciler) reconcileExternal(ctx context.Context, cluster *clus
 		return external.ReconcileOutput{}, err
 	}
 
+	// remove machineset controller onwer, if any since the Machine should be the owner.
+	if controller := metav1.GetControllerOf(obj); controller != nil {
```
Comment (jzhoucliqr) [inline controllers/machine_controller_phases.go:118]: minor nit, but maybe check if controller.Kind == "MachineSet" ?

Comment (srm09) [reply controllers/machine_controller_phases.go:118]: Doesn't the hardcoded value for ownerRef.Kind to `MachineSet` implicitly do the same thing?

Comment (jzhoucliqr) [reply controllers/machine_controller_phases.go:118]: Yeah, it does. If I understand correctly, for performance, without checking Kind it'll always go through all ownerreferences, no matter if the MacineSet is present or not.  Also, for readability, this might be a bit confusing, as we kinds of assume this controller is MachineSet, but in fact it maybe Machine.

Comment (srm09) [reply controllers/machine_controller_phases.go:118]: I still feel adding the check for controller Kind, somehow duplicates the logic. I will defer to the better judgement of experienced folks like @vincepri

File: util/util_test.go [hunk 4]
```diff
@@ -792,3 +792,45 @@ func TestIsSupportedVersionSkew(t *testing.T) {
...
+				Kind:       "Twilight",
+				Name:       "abcdef",
+			},
+			wasRemoved: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			newOwnerRefs := RemoveOwnerRef(ownerRefs, tt.toBeRemoved)
+			g.Expect(len(ownerRefs) != len(newOwnerRefs)).To(Equal(tt.wasRemoved))
```
Comment (fabriziopandini) [inline util/util_test.go:833]: What about using `HasOwnerRef` to check if `toBeRemoved` is gone?

File: controllers/machine_controller_phases.go (hunk 3 above)
Comment (fabriziopandini) [reply controllers/machine_controller_phases.go:118]: My two cents.
I'm +1 to have the check on "controlled by a machineSet" more explicit for improving for readability.

For the same reason, might be worth to split the command in two-line so we will have `ownerRefs := util.RemoveOwnerRef(...)` which makes it clear we are removing from the list, and the `obj.SetOwnerReferences(ownerRefs)` in a separated line

Comment (vincepri) [reply controllers/machine_controller_phases.go:118]: It might be worth adding it for readability, + a comment explaining the reason why this has to be removed this way. In general, owner reference is a very finite set, performance-wise we wouldn't get many improvements.

File: util/util_test.go (hunk 4 above)
Comment (srm09) [reply util/util_test.go]: Makes the test cleaner, good point

**Comment by fabriziopandini** [comment]
LGTM pending implementation of the tests in machine_controller_phases_test.go

**Comment by vincepri** [comment]
@srm09 code changes look good, we only need to get tests in place

File: controllers/machine_controller_phases.go [hunk 5]
```diff
@@ -114,6 +114,19 @@ func (r *MachineReconciler) reconcileExternal(ctx context.Context, cluster *clus
 		return external.ReconcileOutput{}, err
 	}
 
+	// With the migration from v1alpha2 to v1alpha3, Machine controllers should be the owner for the
+	// infra Machines, hence remove any existing machineset controller owner reference
+	if controller := metav1.GetControllerOf(obj); controller != nil && controller.Kind == "MachineSet" {
+		ownerRefs := util.RemoveOwnerRef(obj.GetOwnerReferences(), metav1.OwnerReference{
+			APIVersion: clusterv1.GroupVersion.Version,
```
Comment (sedefsavas) [inline controllers/machine_controller_phases.go:121]: This should be just `clusterv1.GroupVersion.String()`. 
ApiVersion has both group and version info: `cluster.x-k8s.io/v1alpha3`

**Comment by vincepri** [comment]
@srm09 What's the status of this PR?

File: controllers/machine_controller_phases.go [hunk 6]
```diff
@@ -114,6 +114,19 @@ func (r *MachineReconciler) reconcileExternal(ctx context.Context, cluster *clus
...
+			Kind:       "MachineSet",
+			Name:       controller.Name,
+			UID:        controller.UID,
+			Controller: pointer.BoolPtr(true),
```
Comment (detiber) [inline controllers/machine_controller_phases.go:125]: These look to be ignored by the comparisons and should be able to be omitted
```suggestion
```

File: controllers/machine_controller_phases_test.go [hunk 7]
```diff
@@ -752,6 +752,12 @@ func TestReconcileBootstrap(t *testing.T) {
 	}
 }
 
+var _ = Describe("Reconcile external", func() {
+	It("updates the owner references for existing machines managed by Machine Sets", func() {
+
+	})
+})
```
[the remark this answers is not in the document]
Comment (srm09) [reply controllers/machine_controller_phases_test.go]: Removed this and added the tests to existing suites.

File: controllers/machine_controller_phases.go (hunk 5 above)
Comment (srm09) [reply controllers/machine_controller_phases.go]: Makes sense, done

**Comment by srm09** [comment]
> @srm09 What's the status of this PR?

@vincepri I added the tests for the machine_controller_phases with the help of @sedefsavas and also incorporated the suggestions from @detiber

**Comment by fabriziopandini** [comment]
@srm09 great work!
/lgtm

File: controllers/suite_test.go [hunk 8]
```diff
@@ -116,3 +119,36 @@ var _ = AfterSuite(func() {
 		Expect(testEnv.Stop()).To(Succeed())
 	}
 })
+
+func ContainRefOfKind(kind string) types.GomegaMatcher {
```
Comment (vincepri) [inline controllers/suite_test.go:123]: Nice!

File: controllers/machine_controller_phases.go [hunk 9]
```diff
@@ -115,6 +115,17 @@ func (r *MachineReconciler) reconcileExternal(ctx context.Context, cluster *clus
 		return external.ReconcileOutput{}, err
 	}
 
+	// With the migration from v1alpha2 to v1alpha3, Machine controllers should be the owner for the
+	// infra Machines, hence remove any existing machineset controller owner reference
+	if controller := metav1.GetControllerOf(obj); controller != nil && controller.Kind == "MachineSet" {
```
Comment (ncdc) [inline controllers/machine_controller_phases.go:120]: It would probably be safest to check both controller.Kind and the group portion of controller.APIVersion

File: controllers/suite_test.go (hunk 8 above)
Comment (ncdc) [inline controllers/suite_test.go:123]: This should probably be ContainRefOfGroupKind and take in a group and a kind. (Also this is awesome, btw!)

File: controllers/machine_controller_phases.go [hunk 10]
```diff
@@ -115,6 +116,23 @@ func (r *MachineReconciler) reconcileExternal(ctx context.Context, cluster *clus
...
+		gv, err := schema.ParseGroupVersion(controller.APIVersion)
+		if err != nil {
+			return external.ReconcileOutput{}, err
+		}
+		if gv.Group == clusterv1.GroupVersion.Group {
+			ownerRefs := util.RemoveOwnerRef(obj.GetOwnerReferences(), metav1.OwnerReference{
+				APIVersion: clusterv1.GroupVersion.String(),
```
Comment (sedefsavas) [inline controllers/machine_controller_phases.go:128]: Crit. If previous MachineSet owner has v1alpha2 version, does this not fail to remove it?
Can we add this scenario to the tests if it is not there already?

Comment (srm09) [reply controllers/machine_controller_phases.go:128]: We are just checking the Group portion of the APIVersion. I can add a separate test for the negative scenario here.

Comment (vincepri) [reply controllers/machine_controller_phases.go:128]: That's a good point, can we just pass in the `controller` variable in here instead of creating a new OwnerReference? Given that we're checking the Kind and Group now, it should be safe.

File: controllers/suite_test.go [hunk 11]
```diff
@@ -116,3 +120,42 @@ var _ = AfterSuite(func() {
...
+		if ref.Kind == matcher.kind && gv.Group == clusterv1.GroupVersion.Group {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func (matcher *refKindMatcher) FailureMessage(actual interface{}) (message string) {
+	return fmt.Sprintf("Expected %+v to contain refs of Kind %s", actual, matcher.kind)
```
Comment (ncdc) [inline controllers/suite_test.go:156]: Please update to include the group

File: controllers/suite_test.go [hunk 12]
```diff
@@ -116,3 +120,42 @@ var _ = AfterSuite(func() {
...
+}
+
+func (matcher *refKindMatcher) NegatedFailureMessage(actual interface{}) (message string) {
+	return fmt.Sprintf("Expected %+v not to contain refs of Kind %s", actual, matcher.kind)
```
Comment (ncdc) [inline controllers/suite_test.go:160]: Please update to include the group

File: controllers/suite_test.go [hunk 13]
```diff
@@ -116,3 +120,42 @@ var _ = AfterSuite(func() {
...
 })
+
+func ContainRefOfGroupKind(group, kind string) types.GomegaMatcher {
+	return &refKindMatcher{
+		kind:  kind,
+		group: group,
+	}
+}
+
+type refKindMatcher struct {
```
Comment (ncdc) [inline controllers/suite_test.go:131]: Please rename to something like refGroupKindMatcher

File: controllers/suite_test.go [hunk 14]
```diff
@@ -116,3 +120,42 @@ var _ = AfterSuite(func() {
...
+	kind  string
+	group string
+}
+
+func (matcher *refKindMatcher) Match(actual interface{}) (success bool, err error) {
+	ownerRefs, ok := actual.([]metav1.OwnerReference)
+	if !ok {
+		return false, fmt.Errorf("ContainRefOfKind matcher expects a slice of OwnerReference")
```
Comment (ncdc) [inline controllers/suite_test.go:139]:
```suggestion
		return false, errors.Errorf("expected []metav1.OwnerReference; got %T", actual)
```

**Comment by vincepri** [comment]
@srm09 Do you have time to address the comments above?

**Comment by srm09** [comment]
@vincepri I have resolved all the review comments.

File: controllers/machine_controller_phases_test.go [hunk 15]
```diff
@@ -681,13 +681,22 @@ func TestReconcileBootstrap(t *testing.T) {
 			},
 		},
 		{
-			name: "existing machine, bootstrap provider is not ready",
+			name: "existing machine, bootstrap provider is not ready, and ownerref updated",
```
Comment (ncdc) [inline controllers/machine_controller_phases_test.go:684]: Sorry to think of this at the last minute, but could you please add a **new** test case that hard-codes the v1alpha2 API version for a MachineSet owner, just to cover all our bases? Thanks!

File: controllers/suite_test.go (hunk 13 above)
Comment (ncdc) [reply controllers/suite_test.go:134]: Bump

File: controllers/suite_test.go (hunk 11 above)
Comment (ncdc) [reply controllers/suite_test.go:159]: Bump

File: controllers/suite_test.go (hunk 12 above)
Comment (ncdc) [reply controllers/suite_test.go:163]: Bump

**Comment by ncdc** [comment]
@srm09 just a couple of small issues to resolve and then we're good to merge. Thanks for your patience with the reviews!

File: controllers/suite_test.go (hunk 13 above)
Comment (srm09) [reply controllers/suite_test.go:134]: I did an incorrect rebase and lost that commit, adding it back
